cache_test.go

```diff
@@ -656,3 +656,89 @@ func (s sizedString) Size() int { return len(s) }
 func (s sizedString) MarshalBinary() (data []byte, err error) {
 	return []byte(s), nil
 }
+
+type mockPubSub struct {
+	calledKeys []string
+	fns        []func(fromID string, key string)
+	sync.Mutex
+	sync.WaitGroup
+}
+
+func (m *mockPubSub) CalledKeys() []string {
+	m.Lock()
+	defer m.Unlock()
+	return m.calledKeys
+}
+
+func (m *mockPubSub) Subscribe(fn func(fromID string, key string)) error {
+	m.Lock()
+	defer m.Unlock()
+	m.fns = append(m.fns, fn)
+	return nil
+}
+
+func (m *mockPubSub) Publish(fromID, key string) error {
+	m.Lock()
+	defer m.Unlock()
+	m.calledKeys = append(m.calledKeys, key)
+	for _, fn := range m.fns {
+		fn := fn
+		m.Add(1)
+		// run in goroutine to prevent deadlock
+		go func() {
+			fn(fromID, key)
+			m.Done()
+		}()
+	}
+	return nil
+}
+
+func newRedisPubSub(channel string) *redisPubSub {
```

Comment:
this redis implementation of pubsub better to be real i.e. moved to the code. We added a dependency to redis, so let's do it for the code, not just a test.